tests/tracer/test_tracer.py

```diff
@@ -395,71 +392,66 @@ def test_start_span_service_from_parent(self):
     def test_start_span_service_global_config(self):
         # When no service is provided a default
         with self.override_global_config(dict(service="mysvc")):
-            span = self.start_span("")
-            span.assert_matches(service="mysvc")
+            with self.start_span("") as span:
+                span.assert_matches(service="mysvc")
 
     def test_start_span_service_global_config_parent(self):
         # Parent should have precedence over global config
         with self.override_global_config(dict(service="mysvc")):
             with self.start_span("parent", service="parentsvc") as parent:
-                child = self.start_span("child", child_of=parent)
-
+                with self.start_span("child", child_of=parent) as child:
+                    pass
         child.assert_matches(
             name="child",
             service="parentsvc",
         )
 
     def test_start_child_span(self):
         # it should create a child Span for the given parent
-        parent = self.start_span("web.request")
-        child = self.start_span("web.worker", child_of=parent)
+        with self.start_span("web.request") as parent:
+            assert self.tracer.current_span() is None
```

Comment:
I'm a little confused by this but I guess it's expected.
As in, if you start a span why the `current_span` is `None`?

Reply:
In general though it's expected behaviour that creating a span doesn't necessarily "activate" it -- hence the `tracer.trace()` API. However, this isn't very well documented which it definitely should be. I added an `activate` flag to start_span in #1936 to hopefully make this more explicit.

What is actually happening behind the scenes is that no parent is given to start_span so this runs https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/tracer.py#L398-L400

Which creates a new context specifically for this new span. That context isn't stored with the context management so `current_span()` won't find it since it looks for the context from the context management... I don't think I'm a fan of how implicit that is 😆

Reply:
Is this behavior documented somewhere? Would be good to make sure when we look at code like this in a few months we remember why it is this way.

Reply:
Yes to improving the API and documenting it.
This has been my pain point since day 1 working on this, and as you can see, it's still not obvious to an "insider".